File: src/main/resources/public/ts/services/QuestionChoiceService.ts
import {idiom, ng, notify} from 'entcore';
import http from 'axios';
import {QuestionChoice} from '../models';
import {DataUtils} from "../utils";

export interface QuestionChoiceService {
    list(questionId: number) : Promise<any>;
    listChoices(questionIds) : Promise<any>;
    get(choiceId: number) : Promise<any>;
    save(choice: QuestionChoice) : Promise<any>;
    create(choice: QuestionChoice) : Promise<any>;
    createMultiple(choices: QuestionChoice[], questionId: number) : Promise<any>;
    update(choice: QuestionChoice) : Promise<any>;
    delete(choiceId: number) : Promise<any>;
}

export const questionChoiceService: QuestionChoiceService = {

    async list(questionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/questions/${questionId}/choices`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionChoiceService.list'));
            throw err;
        }
    },

    async listChoices(questionIds) : Promise<any>{
        try{
            return DataUtils.getData(await http.get(`/formulaire/questions/choices/all`, { params: questionIds }));
        } catch(err){
            notify.error(idiom.translate('formulaire.error.questionChoiceService.list'));
            throw err;
        }
    },

    async get(choiceId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/choices/${choiceId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionChoiceService.get'));
            throw err;
        }
    },

    async save(choice: QuestionChoice) : Promise<any> {
        return choice.id ? await this.update(choice) : await this.create(choice);
    },

    async create(choice: QuestionChoice) : Promise<any> {
        try {
            return DataUtils.getData(await http.post(`/formulaire/questions/${choice.question_id}/choices`, choice));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionChoiceService.create'));
            throw err;
        }
    },

    async createMultiple(choices: QuestionChoice[], questionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.post(`/formulaire/questions/${questionId}/choices/multiple`, choices));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionChoiceService.createMultiple'));
            throw err;
        }
    },

    async update(choice: QuestionChoice) : Promise<any> {
        try {
            return DataUtils.getData(await http.put(`/formulaire/choices/${choice.id}`, choice));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionChoiceService.update'));
            throw err;
        }
    },

    async delete(choiceId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.delete(`/formulaire/choices/${choiceId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.questionChoiceService.delete'));
            throw err;
        }
    }
};

export const QuestionChoiceService = ng.service('QuestionChoiceService',(): QuestionChoiceService => questionChoiceService);

File: formulaire/src/main/java/fr/openent/formulaire/service/QuestionChoiceService.java
package fr.openent.formulaire.service;

import fr.openent.form.core.models.QuestionChoice;
import fr.wseduc.webutils.Either;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.List;

public interface QuestionChoiceService {
    /**
     * List all the choices of a specific question
     * @param questionId question identifier
     * @param handler function handler returning JsonArray data
     */
    void list(String questionId, Handler<Either<String, JsonArray>> handler);

    /**
     * ListChoices all the choices of a specific question
     * @param questionIds JsonArray identifier
     * @param handler function handler returning JsonArray data
     */
    void listChoices(JsonArray questionIds, Handler<Either<String, JsonArray>> handler);

    /**
     * Get a specific by id
     * @param choiceId choice identifier
     * @param handler function handler returning JsonObject data
     */
    void get(String choiceId, Handler<Either<String, JsonObject>> handler);

    /**
     * Create a choice for a specific question
     * @param questionId question identifier
     * @param choice QuestionChoice data
     * @param locale locale language
     */
    Future<JsonObject> create(String questionId, QuestionChoice choice, String locale);

    /**
     * Duplicate choices of a specific question
     * @param formId new form identifier
     * @param questionId question identifier
     * @param originalQuestionId original question identifier
     * @param handler function handler returning JsonObject data
     */
    void duplicate(int formId, int questionId, int originalQuestionId, Handler<Either<String, JsonObject>> handler);

    /**
     * Update a specific choice
     * @param choice QuestionChoice data
     * @param locale locale language
     */
    Future<JsonObject> update(QuestionChoice choice, String locale);

    /**
     * Update a list of specific choices
     * @param choices List<QuestionChoice> data
     * @param locale locale language
     */
    Future<JsonArray> update(List<QuestionChoice> choices, String locale);

    /**
     * Delete a specific choice
     * @param choiceId choice identifier
     */
    Future<JsonObject> delete(String choiceId);

    /**
     * Check validity of a specific choice
     * @param choice choice to check
     */
    Future<Boolean> isTargetValid(QuestionChoice choice);
}